### PR TITLE
Added kaiming_uniform initialization for Conv2d and Linear layers

### PR DESCRIPTION
**Pull request by rabiaedayilmaz**
Added Kaiming Uniform Initialization as the default.

It is the default method, aka He weight initialization, in Pytorch:
https://discuss.pytorch.org/t/whats-the-default-initialization-methods-for-layers/3157/21?page=2

### CODE REVIEW

File: tinygrad/nn/__init__.py [hunk 1]
```diff
@@ -34,19 +34,20 @@ def __call__(self, x:Tensor):
...
 
 # TODO: is this good weight init?
+# kaiming initialization is added as default
 class Conv2d:
-  def __init__(self, in_channels, out_channels, kernel_size, stride=1, padding=0, dilation=1, groups=1, bias=True):
+  def __init__(self, in_channels, out_channels, kernel_size, stride=1, padding=0, dilation=1, groups=1, bias=True, initialization: str='kaiming_uniform'):
     self.kernel_size = (kernel_size, kernel_size) if isinstance(kernel_size, int) else (kernel_size[0], kernel_size[1])
     self.stride, self.padding, self.dilation, self.groups = stride, padding, dilation, groups
-    self.weight = Tensor.glorot_uniform(out_channels, in_channels//groups, self.kernel_size[0], self.kernel_size[1])
+    self.weight = Tensor.kaiming_uniform(out_channels, in_channels//groups, self.kernel_size[0], self.kernel_size[1]) if initialization=='kaiming_uniform' else Tensor.glorot_uniform(out_channels, in_channels//groups, self.kernel_size[0], self.kernel_size[1])
```
Comment (geohot) [inline tinygrad/nn/__init__.py:42]: getattr(Tensor, initialization)?

Comment (rabiaedayilmaz) [reply tinygrad/nn/__init__.py]: Great idea, thank you, changed them to getattr(Tensor, initialization) version

Comment (rabiaedayilmaz) [reply tinygrad/nn/__init__.py]: and fixed some silly typos (:

File: tinygrad/nn/__init__.py [hunk 2]
```diff
@@ -34,19 +34,21 @@ def __call__(self, x:Tensor):
...
 
 # TODO: is this good weight init?
+# kaiming initialization is added as default
 class Conv2d:
-  def __init__(self, in_channels, out_channels, kernel_size, stride=1, padding=0, dilation=1, groups=1, bias=True):
+  def __init__(self, in_channels, out_channels, kernel_size, stride=1, padding=0, dilation=1, groups=1, bias=True, initialization: str='kaiming_uniform'):
     self.kernel_size = (kernel_size, kernel_size) if isinstance(kernel_size, int) else (kernel_size[0], kernel_size[1])
     self.stride, self.padding, self.dilation, self.groups = stride, padding, dilation, groups
-    self.weight = Tensor.glorot_uniform(out_channels, in_channels//groups, self.kernel_size[0], self.kernel_size[1])
+    self.weight = getattr(Tensor, initialization)(out_channels, in_channels // groups, self.kernel_size[0],
```
Comment (geohot) [inline tinygrad/nn/__init__.py:42]: use one line here

**Comment by geohot** [comment]
Before merging this, I want some statistical tests that it matches torch. Or even better if you can set the seeds and get a perfect match.

**Comment by jla524** [comment]
I've added some tests to check if the distributions match torch.
We can add something like this in `test/test_randomness.py`.
```
def test_kaiming_uniform(self):
  self.assertFalse(normal_test(Tensor.glorot_uniform))
  self.assertTrue(equal_distribution(Tensor.kaiming_uniform, lambda x: torch.nn.init.kaiming_uniform_(torch.empty(x)), lambda x: (np.random.rand(*x) * 2 - 1) * 3 / x[0]))
```
Currently it fails, as the distribution doesn't match torch.
I wonder if it has something to do with [nonlinearity in torch](https://pytorch.org/docs/stable/_modules/torch/nn/init.html#kaiming_normal_).

**Comment by rabiaedayilmaz** [comment]
I set seed and made local investigations to my last commit. I found that the means of the Conv2D layer torch and Tensor outputs are statistically similar according to` scipy.stats.ttest_ind` function and the standard deviations of the Conv2D layer torch and Tensor outputs are different by being 1.3293099403381348 and 1.3305704593658447, an error of 1e-2.

**Comment by rabiaedayilmaz** [comment]
Do we require any additional tests? @geohot @jla524

**Comment by rabiaedayilmaz** [comment]
Oh okay, got it. Tested with torch and implemented kaiming_uniform functions and they match. Is there anything that should be added? @geohot

File: tinygrad/tensor.py [hunk 3]
```diff
@@ -162,8 +162,54 @@ def scaled_uniform(*shape, **kwargs) -> Tensor: return Tensor.uniform(*shape, **
...
+      if dims > 2:
+        for s in shape[2:]:
+          receptive_field_size *= s
+
+      fan_in = num_input_fmaps * receptive_field_size
+      fan_out = num_output_fmaps * receptive_field_size
+
+      return fan_in if mode == "fan_in" else fan_out
+
+    def _calculate_gain(nonlinearity, a):
```
Comment (geohot) [inline tinygrad/tensor.py:186]: This is a brutal number of lines.

Comment (rabiaedayilmaz) [reply tinygrad/tensor.py]: And adjusted test func.

File: tinygrad/tensor.py [hunk 4]
```diff
@@ -162,8 +162,14 @@ def scaled_uniform(*shape, **kwargs) -> Tensor: return Tensor.uniform(*shape, **
   @staticmethod
   def glorot_uniform(*shape, **kwargs) -> Tensor: return Tensor.uniform(*shape, **kwargs).mul((6/(shape[0]+prod(shape[1:])))**0.5)
 
-  # ***** toposort and backward pass *****
+  # ***** kaiming uniform to weights *****
+  # https://pytorch.org/docs/stable/_modules/torch/nn/init.html#kaiming_uniform_
+  @staticmethod
+  def kaiming_uniform(*shape, a:float = 0.01, **kwargs) -> Tensor:
+    bound = math.sqrt(3.0) * math.sqrt(2.0 / (1 + a ** 2)) / math.sqrt(shape[1] * prod(shape[2:]))
+    return Tensor(LazyNumpyArray(lambda lna: np.random.uniform(low=-bound, high=bound, size=lna.shape), shape, np.float32))
```
Comment (geohot) [inline tinygrad/tensor.py:170]: Use `Tensor.uniform` like the above init functions.

Comment (rabiaedayilmaz) [reply tinygrad/tensor.py:170]: But `Tensor.uniform` does not have bound parameters for low and high bounds. What should I do?

Comment (rabiaedayilmaz) [reply tinygrad/tensor.py]: Hey, @geohot, I updated `Tensor.uniform` since it wasn't generalizable to a specific range, according to [1]. So, I was able to change `return` line. It is an unpalatable practice to update two functions, although it works fine, in a pull request, but I had to do it.

[1] https://numpy.org/doc/stable/reference/random/generated/numpy.random.Generator.uniform.html#numpy.random.Generator.uniform

File: tinygrad/tensor.py [hunk 5]
```diff
@@ -158,7 +158,7 @@ def randn(*shape, **kwargs) -> Tensor: return Tensor(LazyNumpyArray(lambda lna:
   # ***** rng hlops *****
 
   @staticmethod
-  def uniform(*shape, **kwargs) -> Tensor: return Tensor.rand(*shape, **kwargs) * 2 - 1
+  def uniform(*shape, low=-1.0, high=1.0, **kwargs) -> Tensor: return Tensor(LazyNumpyArray(lambda lna: Tensor._rng.uniform(low=low, high=high, size=lna.shape), shape, np.float32), **kwargs)
```
Comment (geohot) [inline tinygrad/tensor.py:161]: Should be good after one more change.

I'm fine with the low and high params, but use Tensor.rand as a base, not another LazyNumpyArray.

Comment (rabiaedayilmaz) [reply tinygrad/tensor.py:161]: I see but I think I shouldn't. Because to use `Tensor.rand` as a base, I need to change that function from `Tensor._rng.random` to `Tensor._rng.uniform`. Is it okay?

Comment (geohot) [reply tinygrad/tensor.py:161]: Please look into that more.

Comment (rabiaedayilmaz) [reply tinygrad/tensor.py:161]: Hm, because `np.random.Generator.random` does not have low and high params but `np.random.Generator.uniform` does, it is better to use LazyNumpyArray? Am I missing something?

Comment (rabiaedayilmaz) [reply tinygrad/tensor.py]: Done! ^_^

File: tinygrad/tensor.py [hunk 6]
```diff
@@ -164,7 +164,7 @@ def randn(*shape, **kwargs) -> Tensor: return Tensor(LazyNumpyArray(lambda lna:
   # ***** rng hlops *****
 
   @staticmethod
-  def uniform(*shape, **kwargs) -> Tensor: return Tensor.rand(*shape, **kwargs) * 2 - 1
+  def uniform(*shape, low=-1.0, high=1.0, **kwargs) -> Tensor: return ((high-low) * Tensor.rand(*shape, **kwargs)) + low
```
Comment (geohot) [inline tinygrad/tensor.py:167]: Nit: This is right, but why did the math order change?

**Comment by geohot** [comment]
Merged! Always think about what the absolute minimum change is to accomplish a task, that's the tinygrad way.
